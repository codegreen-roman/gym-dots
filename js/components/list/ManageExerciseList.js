import React, { Component } from 'react'
import { ExerciseList } from './ExerciseList'
import { connect } from 'react-redux'
import { compose } from 'ramda'
import { object, func } from 'prop-types'
import { fetchExercises, exercisesOrderChange } from '../../state/actions'

class _ManageExerciseList extends Component {
    constructor(props) {
        super(props)
    }

    componentWillMount() {
        this.props.fetchExercises()
    }

    render() {
        const { exercises: { upcoming, completed, skipped }, onOrderChange } = this.props
        return (
            <div>
                <ExerciseList exercises={upcoming} title='upcoming' onOrderChangeClick={onOrderChange} />
                <ExerciseList exercises={completed} title='completed' />
                <ExerciseList exercises={skipped} title='skipped' />
            </div>
        )
    }
}

_ManageExerciseList.propTypes = {
    exercises: object.isRequired,
    fetchExercises: func.isRequired,
    onOrderChange: func.isRequired
}

const mapStateToProps = ({ exercises, sessionId, name }) => {
    return {
        exercises,
        sessionId,
        name
    }
}

const mapActionsToProps = dispatch => {
    return {
        fetchExercises: compose(dispatch, fetchExercises),
        onOrderChange: compose(dispatch, exercisesOrderChange)
    }
}

export { _ManageExerciseList }
export const ManageExerciseList = connect(mapStateToProps, mapActionsToProps)(_ManageExerciseList)
