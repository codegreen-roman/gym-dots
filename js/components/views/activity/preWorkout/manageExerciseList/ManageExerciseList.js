import React, { Component } from 'react'
import { ExerciseList } from './ExerciseList'
import { connect } from 'react-redux'
import { compose, all, isEmpty, not, isNil } from 'ramda'
import { func, bool, string, array } from 'prop-types'
import { exercisesOrderChange } from '../../../../../state/actions/index'

const notMissing = compose(not, isNil)
const allListsEmpty = all(isEmpty)

class _ManageExerciseList extends Component {

    congratulateUser() {
        const { sessionDone, sessionKey } = this.props

        if (sessionDone) {
            return (<h2>Your session {sessionKey} is COMPLETED!</h2>)
        }

        return null
    }

    render() {
        const { upcoming, completed, skipped, onOrderChange, } = this.props

        return (
            <div>
                {this.congratulateUser()}
                <span>{isEmpty(upcoming) ? 'You got no exercises' : null}</span>
                <ExerciseList key='upcoming' list={upcoming} title='upcoming' onOrderChangeClick={onOrderChange} />
                <ExerciseList key='completed' list={completed} title='completed' />
                <ExerciseList key='skipped' list={skipped} title='skipped' />
            </div>
        )
    }
}

_ManageExerciseList.propTypes = {
    upcoming: array.isRequired,
    completed: array.isRequired,
    skipped: array.isRequired,
    onOrderChange: func.isRequired,
    sessionKey: string,
    sessionDone: bool.isRequired
}

const mapStateToProps = ({ exercises: { upcoming, completed, skipped, sessionKey, name } }) => ({
    upcoming,
    completed,
    skipped,
    name,
    sessionKey,
    sessionDone: allListsEmpty([upcoming, skipped]) && notMissing(sessionKey)
})

const mapActionsToProps = dispatch => {
    return {
        onOrderChange: compose(dispatch, exercisesOrderChange)
    }
}

export { _ManageExerciseList }
export const ManageExerciseList = connect(mapStateToProps, mapActionsToProps)(_ManageExerciseList)
