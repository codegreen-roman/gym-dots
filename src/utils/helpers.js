import {
    curry,
    compose,
    not,
    isEmpty,
    addIndex,
    map,
    prop
} from 'ramda'
/**
 * branch
 *
 * this function accepts a predicate function
 * and returns the first component if true, and the second
 * if predicate is false
 *
 * branch :: (Function -> Bool) -> Element -> Element -> Elment
 *
 * @param {function|boolean} [predicatefn] function that returns a boolean or a boolean
 * @param {node}             [comp1]       element to render if predicate is true
 * @param {node}             [comp2]       element to render if predicate is false
 * @returns {node}                         returns the correct component according to predicate
 */
export const branch = curry((predicatefn, comp1, comp2) => (predicatefn ? comp1 : comp2)); // eslint-disable-line

/**
 * Function (Component) that returns null.
 * Example usage: branch(isEmpty(items), <RenderNothing />, <Item items={items} />)
 */
export const RenderNothing = () => null

/**
 * Opposite of isEmpty
 */
export const notEmpty = compose(not, isEmpty)
// tried const notEmpty = complement(isEmpty), which is the same, but test ManageExerciseList.helper.spec.js:8 failed =(

/**
 * Adding indexes to map
 */
export const mapIndexed = addIndex(map)

/**
 * Exctracting event.target.value
 */
export const extractTargetValue = compose(prop('value'), prop('target'))
