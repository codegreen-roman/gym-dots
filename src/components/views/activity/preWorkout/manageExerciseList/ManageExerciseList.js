import React, { Component } from 'react'
import { ExerciseList } from './ExerciseList'
import { connect } from 'react-redux'
import { compose, isEmpty } from 'ramda'
import { toWritableResults, allListsEmpty, notMissing } from './ManageExerciseList.helper'
import { func, bool, string, array } from 'prop-types'
import { exercisesOrderChange, saveExercisesResults } from '../../../../../state/actions'

class _ManageExerciseList extends Component {

    componentDidMount() {
        const { sessionDone, sessionKey, completed, saveResults } = this.props
        if (sessionDone) {

            saveResults({
                [sessionKey]: toWritableResults(completed)
            })
        }
    }

    congratulateUser() {
        const { sessionDone, sessionKey } = this.props

        if (sessionDone) {
            return (<h2>Your session {sessionKey} is COMPLETED!</h2>)
        }

        return null
    }

    render() {
        const { upcoming, completed, skipped, onOrderChange, } = this.props

        return (
            <div>
                {this.congratulateUser()}
                <span>{isEmpty(upcoming) ? 'You got no exercises yet' : null}</span>
                <ExerciseList key='upcoming' list={upcoming} title='upcoming' onOrderChangeClick={onOrderChange} />
                <ExerciseList key='completed' list={completed} title='completed' />
                <ExerciseList key='skipped' list={skipped} title='skipped' />
            </div>
        )
    }
}

_ManageExerciseList.propTypes = {
    upcoming: array.isRequired,
    completed: array.isRequired,
    skipped: array.isRequired,
    onOrderChange: func.isRequired,
    saveResults: func.isRequired,
    sessionKey: string,
    sessionDone: bool.isRequired
}

const mapStateToProps = ({ exercises: { upcoming, completed, skipped, sessionKey, name } }) => ({
    upcoming,
    completed,
    skipped,
    name,
    sessionKey,
    sessionDone: allListsEmpty([upcoming, skipped]) && notMissing(sessionKey)
})

const mapActionsToProps = dispatch => {
    return {
        saveResults: compose(dispatch, saveExercisesResults),
        onOrderChange: compose(dispatch, exercisesOrderChange)
    }
}

export { _ManageExerciseList }
export const ManageExerciseList = connect(mapStateToProps, mapActionsToProps)(_ManageExerciseList)
