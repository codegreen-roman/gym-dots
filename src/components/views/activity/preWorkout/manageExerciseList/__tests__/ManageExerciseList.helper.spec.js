import { toWritableResults, allListsEmpty } from '../ManageExerciseList.helper'
import { notEmpty } from '@utils/helpers'

test('.notMissing returns true or false for strings', () => {
    expect(typeof notEmpty === 'function').toBeTruthy()
    expect(notEmpty('12345')).toBeTruthy()
    expect(notEmpty('')).toBeFalsy()
    expect(notEmpty()).toBeFalsy()
})

test('.allListsEmpty returns true for a list of empty lists', () => {
    expect(typeof allListsEmpty === 'function').toBeTruthy()
    expect(allListsEmpty([[], [], []])).toBeTruthy()
})

test('.allListsEmpty returns false for a list of empty lists and not empty list', () => {
    expect(allListsEmpty([[], [], [1]])).toBeFalsy()
})

test('.toWritableResults returns a freaking structure eatable by firebase', () => {

    const completed = [
        {
            exerciseKey: 'key1',
            restTime: 90,
            name: 'Push-ups',
            sets: 5,
            reps: 20,
            weight: 0,
            allDone: true
        },
        {
            exerciseKey: 'key2',
            restTime: 30,
            name: 'Australian pull-ups',
            sets: 5,
            reps: 12,
            weight: 0,
            allDone: false
        },
        {
            exerciseKey: 'key3',
            restTime: 30,
            name: 'Australian pull-ups',
            sets: 5,
            reps: 12,
            weight: 0,
            allDone: false
        },
        {
            exerciseKey: 'key4',
            restTime: 30,
            name: 'Australian pull-ups',
            sets: 5,
            reps: 12,
            weight: 0,
            allDone: true
        }
    ]

    expect(typeof toWritableResults === 'function').toBeTruthy()
    expect(toWritableResults(completed)).toMatchObject({
        'key1': true,
        'key2': false,
        'key3': false,
        'key4': true
    })
    expect(toWritableResults(completed)).toMatchSnapshot()
})
