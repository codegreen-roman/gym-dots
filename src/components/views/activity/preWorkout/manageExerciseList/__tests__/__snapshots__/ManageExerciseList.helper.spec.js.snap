// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.toWritableResults returns a freaking structure eatable by firebase 1`] = `
Object {
  "key1": true,
  "key2": false,
  "key3": false,
  "key4": true,
}
`;
